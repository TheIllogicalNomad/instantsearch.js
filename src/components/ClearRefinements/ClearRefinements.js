import PropTypes from 'prop-types';
import React, { Component } from 'preact-compat';
import cx from 'classnames';

import Template from '../Template.js';
import { isSpecialClick } from '../../lib/utils.js';

export class ClearRefinements extends Component {
  componentWillMount() {
    this.handleClick = this.handleClick.bind(this);
  }

  shouldComponentUpdate(nextProps) {
    return (
      this.props.url !== nextProps.url ||
      this.props.hasRefinements !== nextProps.hasRefinements
    );
  }

  handleClick(e) {
    if (isSpecialClick(e)) {
      // do not alter the default browser behavior
      // if one special key is down
      return;
    }
    e.preventDefault();
    this.props.refine();
  }

  render() {
    const { hasRefinements, cssClasses } = this.props;
    const data = { hasRefinements };

    const rootClassNames = cx(cssClasses.root);
    const buttonClassNames = cx(cssClasses.button, {
      [cssClasses.disabledButton]: hasRefinements,
    });

    return (
      <div className={rootClassNames}>
        <Template
          data={data}
          templateKey="resetLabel"
          rootTagName="button"
          rootProps={{
            className: buttonClassNames,
            onClick: this.handleClick,
          }}
          {...this.props.templateProps}
        />
      </div>
    );
  }
}

ClearRefinements.propTypes = {
  refine: PropTypes.func.isRequired,
  cssClasses: PropTypes.shape({
    root: PropTypes.string,
    button: PropTypes.string,
    disabledButton: PropTypes.string,
  }),
  hasRefinements: PropTypes.bool.isRequired,
  templateProps: PropTypes.object.isRequired,
  url: PropTypes.string.isRequired,
};

export default ClearRefinements;
