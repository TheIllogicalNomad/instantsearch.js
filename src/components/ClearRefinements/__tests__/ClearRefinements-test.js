import React from 'react';
import sinon from 'sinon';
import ClearRefinements from '../ClearRefinements';
import renderer from 'react-test-renderer';

describe('ClearAll', () => {
  const defaultProps = {
    refine: () => {},
    cssClasses: {
      root: 'root',
      button: 'button',
      disabledButton: 'disabled',
    },
    hasRefinements: false,
    templateProps: {
      templates: {
        resetLabel: '',
      },
    },
    url: '#all-cleared!',
  };

  it('should render <ClearRefinements />', () => {
    const tree = renderer
      .create(<ClearRefinements {...defaultProps} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should handle clicks (and special clicks)', () => {
    const props = {
      refine: sinon.spy(),
    };
    const preventDefault = sinon.spy();
    const component = new ClearRefinements(props);
    ['ctrlKey', 'shiftKey', 'altKey', 'metaKey'].forEach(e => {
      const event = { preventDefault };
      event[e] = true;
      component.handleClick(event);
      expect(props.refine.called).toBe(false, 'refine never called');
      expect(preventDefault.called).toBe(false, 'preventDefault never called');
    });
    component.handleClick({ preventDefault });
    expect(props.refine.calledOnce).toBe(true, 'refine called once');
    expect(preventDefault.calledOnce).toBe(true, 'preventDefault called once');
  });
});
