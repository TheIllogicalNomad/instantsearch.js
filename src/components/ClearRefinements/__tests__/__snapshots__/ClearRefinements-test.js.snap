// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClearAll should render <ClearRefinements /> 1`] = `
<div
  className="root"
>
  <button
    className="button"
    dangerouslySetInnerHTML={
      Object {
        "__html": "",
      }
    }
    onClick={[Function]}
  />
</div>
`;
